### PR TITLE
Foil Shape Similarity Bias: Extend GSB with Inter-Foil Geometry

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Gap/Stagger Spatial Bias (GSB, PR #2130) conditions slice routing on (gap, stagger) scalars — tandem **arrangement** geometry. But it knows nothing about **foil shape similarity**. When two foils are geometrically similar (both NACA0012-like), tandem wake interaction differs fundamentally from a dissimilar pair (NACA0012 fore + NACA6416 aft). The val_tandem_transfer split specifically tests this dissimilar-geometry case.

**The fix:** Extend the GSB's `raw_xy` tensor from 6D to 7D by appending a scalar measuring the shape similarity between foil-1 and foil-2 DSDF representations. This scalar tells the spatial_bias MLP: "these two foils are [similar / different]," enabling geometry-type-conditioned slice routing.

**Why this helps p_tan specifically:** p_tan = val_tandem_transfer = NACA6416 aft foil with unseen shape. With shape-similarity conditioning, the routing network can learn: "low similarity → activate tandem-specialization slices more aggressively; high similarity → rely on standard wake physics." This is exactly the discrimination needed.

**Why it won't hurt:** Zero-initialization of the new input column's weights ensures training starts from the IDENTICAL baseline as the current model. No regression risk on p_in, p_oodc.

**Implementation cost:** ~30 lines. Pure additive.

### Step 1 — Compute inter-foil shape similarity scalar

In the dataset or feature construction code, compute a scalar measuring how similar foil-1 and foil-2 DSDF vectors are. Two options (pick whichever is simpler given the code structure):

**Option A — per-sample cosine similarity of DSDF means:**
```python
# Assume foil-1 DSDF channels: x[:, 2:6], foil-2 DSDF channels: x[:, 6:10]
# x shape: [B, N, F] or [N, F] depending on batching

# Mean over nodes for each DSDF channel
dsdf1_mean = x[..., 2:6].mean(dim=-2)   # [B, 4] or [4]
dsdf2_mean = x[..., 6:10].mean(dim=-2)  # [B, 4] or [4]

# Cosine similarity between the two DSDF vectors
cos_sim = F.cosine_similarity(dsdf1_mean, dsdf2_mean, dim=-1, eps=1e-8)  # [B] or scalar
```

**Option B — L2 distance between DSDF magnitude means (simpler):**
```python
dsdf1_mag = x[..., 2:6].norm(dim=-1).mean(dim=-2)  # [B] or scalar
dsdf2_mag = x[..., 6:10].norm(dim=-1).mean(dim=-2)
shape_diff = (dsdf1_mag - dsdf2_mag).abs()  # absolute difference in mean DSDF magnitude
```

Use whichever approach integrates cleanly with the existing feature pipeline. The key requirement: it's a scalar (or [B, 1]) that can be appended to the existing GSB raw_xy vector.

### Step 2 — Extend GSB raw_xy

Find where `gap_stagger_spatial_bias` builds its `raw_xy` tensor. Currently it constructs a 6D vector from existing spatial coordinates + (gap, stagger). Append the shape similarity scalar:

```python
# Before (6D):
raw_xy = torch.cat([..., gap_scalar, stagger_scalar], dim=-1)  # [B, N, 6]

# After (7D):
shape_sim = cos_sim[:, None, None].expand_as(raw_xy[..., :1])  # broadcast to [B, N, 1]
raw_xy = torch.cat([..., gap_scalar, stagger_scalar, shape_sim], dim=-1)  # [B, N, 7]
```

### Step 3 — Zero-init the new weight column

The GSB spatial_bias MLP takes `raw_xy` as input (6D currently). After extending to 7D, the new weight column must be zero-initialized to preserve training start from baseline:

```python
# Find the first linear layer of the spatial_bias MLP
# Extend its weight matrix: [out_features, 6] → [out_features, 7]
# with the new column initialized to zeros

old_weight = spatial_bias_mlp.fc1.weight.data  # [out_features, 6]
new_col = torch.zeros(old_weight.shape[0], 1, device=old_weight.device)
spatial_bias_mlp.fc1.weight.data = torch.cat([old_weight, new_col], dim=1)

# Similarly extend input_size in the MLP definition from 6 to 7
```

Add a config flag:
```python
gsb_shape_similarity: bool = False   # extend GSB from 6D to 7D with inter-foil shape sim
```

### Step 4 — Run 2 seeds

Use `--wandb_group phase6/foil-shape-similarity-bias`.

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent alphonse \
  --wandb_name "alphonse/fssb-s42" --wandb_group phase6/foil-shape-similarity-bias \
  --gsb_shape_similarity --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73
cd cfd_tandemfoil && python train.py --agent alphonse \
  --wandb_name "alphonse/fssb-s73" --wandb_group phase6/foil-shape-similarity-bias \
  --gsb_shape_similarity --seed 73 \
  [same flags as above]
```

**Do NOT override SENPAI_MAX_EPOCHS or SENPAI_TIMEOUT_MINUTES.**

## What to Report

- Surface MAE: p_in, p_oodc, p_tan, p_re for both seeds + 2-seed mean, with W&B run IDs
- Which similarity metric did you use (cosine similarity or L2 distance)?
- Does the shape_sim scalar vary meaningfully across the dataset? (Log the range/distribution for a few batches)
- Any training instability at epoch 0? (Should be zero if zero-init is correct)

## Baseline

Current single-model baseline (PR #2130, GSB + PCGrad, 2-seed avg):

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| p_in   | 13.05     | < 13.05        |
| p_oodc | 7.70      | < 7.70         |
| **p_tan** | **28.60** | **< 28.60** |
| p_re   | 6.55      | < 6.55         |

W&B baseline runs: `d7l91p0x` (seed 42, p_tan=28.9), `j9btfx09` (seed 73, p_tan=28.3)

```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results — Foil Shape Similarity Bias (GSB 7D)

Implemented cosine similarity of per-sample mean DSDF vectors (foil-1 `x[:,:,2:6]` vs foil-2 `x[:,:,6:10]`), computed before `feature_cross`. Extended GSB spatial_bias MLP from 6D → 7D with zero-initialized 7th column.

### Results (2 seeds)

| Metric | s42 (`yvrpikxi`) | s73 (`j20oel6q`) | **Mean** | Target | Δ |
|--------|-------------------|-------------------|----------|--------|---|
| p_in   | 12.93 | 13.48 | **13.20** | < 13.05 | +0.15 ❌ |
| p_oodc | 8.04  | 7.82  | **7.93**  | < 7.70  | +0.23 ❌ |
| p_tan  | 30.04 | 29.27 | **29.66** | < 28.60 | +1.06 ❌ |
| p_re   | 6.36  | 6.62  | **6.49**  | < 6.55  | -0.06 ✅ |

Both runs hit 180-min wall-clock timeout as expected.

### Implementation Details
- **Similarity metric:** Cosine similarity of node-averaged DSDF vectors: `cos_sim(mean(dsdf1), mean(dsdf2))` → scalar per sample
- **Integration:** Appended as 7th channel to GSB `raw_xy`: `[x, y, curvature, dist, gap, stagger, shape_sim]`
- **Zero-init:** New column's weights in `spatial_bias[0].weight[:, 6]` initialized to zero → training starts identical to baseline
- **torch.compile:** No recompilation warnings, ~5.1 it/s (identical to baseline)

### Analysis

- **p_oodc regressed** (7.93 vs 7.70 target, +3% regression). The shape similarity signal may be interfering with OOD conditioning routing — single-foil samples have zero DSDF for foil-2, so cosine similarity is degenerate (≈0), which the MLP may be misinterpreting.
- **p_tan worse than baseline** (29.66 vs 28.60). The shape similarity scalar is too coarse — a single number averaging DSDF across all nodes loses spatial structure. The dissimilar-geometry discrimination the hypothesis targeted needs node-level, not sample-level, conditioning.
- **p_re slightly beats target** — likely noise given the other regressions.
- **Seed variance is high** for p_oodc: 8.04 (s42) vs 7.82 (s73), suggesting the feature is unstable.

### Suggested follow-ups

1. **Node-level shape difference** instead of sample-level: use `|dsdf1 - dsdf2|` per-node as additional spatial_bias input (keeps spatial structure).
2. **Separate the single-foil case:** For single-foil samples where foil-2 DSDF ≈ 0, the cosine similarity is meaningless. A learned single-vs-tandem gate before the similarity signal might help.
3. **Abandon sample-level similarity, try per-node foil-2 proximity conditioning:** The node's distance to foil-2 (already available as `raw_dsdf[:,:,4:8].abs().min()`) might be a more informative routing signal than global shape similarity.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2157 because ADVISOR: Closing — foil shape similarity bias regressed all key metrics.

**Results:** p_tan=29.66 (+3.7%), p_oodc=7.93 (+3.0%), p_in=13.20 (+1.1%). Only p_re marginally better.

**Analysis:** The sample-level cosine similarity is too coarse — a single scalar averaging DSDF across all nodes loses the spatial structure needed for p_tan discrimination. The degenerate case for single-foil samples (foil-2 DSDF ≈ 0 → cosine sim meaningless) likely confuses the routing MLP.

**Good insight from your follow-up #3:** Per-node foil-2 proximity conditioning is the right direction, and askeladd is already testing that approach in #2162 (tandem cross-DSDF features with per-node dist_ratio + rel_angle). That takes the node-level approach you correctly identified as necessary.

Moving you to a physics-informed experiment next..
